lib/src/command/add.dart

```diff
@@ -349,37 +432,137 @@ class AddCommand extends PubCommand {
+  }
+
+  /// Parse [package] to return the corresponding [_ParseResult].
+  ///
+  /// [package] must be written in the format
+  /// `dev:<package-name>[:descriptor>]`, where quotations should be used if
+  /// necessary.
+  ///
+  /// `descriptor` is what you would put in a pubspec.yaml in the dependencies
+  /// section.
+  ///
+  /// Assumes that none of '--git-url', '--git-ref', '--git-path', '--path' and
+  /// '--sdk' are present in [argResults], but will obey '--dev'.
+  ///
+  ///
+  /// Examples:
+  /// ```
+  /// retry
+  /// retry:2.0.0
+  /// dev:retry:^2.0.0
+  /// retry:'>=2.0.0'
+  /// retry:'>2.0.0 <3.0.1'
+  /// 'retry:>2.0.0 <3.0.1'
+  /// retry:any
+  /// 'retry:{"path":"../foo"}'
+  /// 'retry:{"git":{"url":"../foo","ref":"branchname"},"version":"^1.2.3"}'
+  /// 'retry:{"sdk":"flutter"}'
+  /// 'retry:{"hosted":"mypub.dev"}'
+  /// ```
+  ///
+  /// The --path --sdk and --git-<option> arguments cannot be combined with a
+  /// non-string descriptor.
+  ///
+  /// If a version constraint is provided when the `--path` or any of the
+  /// `--git-<option>` options are used, a [PackageParseError] will be thrown.
+  ///
+  /// Packages must either be a git, hosted, sdk, or path package. Mixing of
+  /// options is not allowed and will cause a [PackageParseError] to be thrown.
+  ///
+  /// If any of the other git options are defined when `--git-url` is not
+  /// defined, an error will be thrown.
+  ///
+  /// Returns a `ref` of `null` if the descriptor did not specify a source.
+  /// Then the source will be determined by the old-style arguments.
+  _PartialParseResult _parsePackage(String package, ArgResults argResults) {
+    final match =
+        RegExp(r'^(?<name>[^:]*)(?<descriptor>:.*)?$').firstMatch(package);
+    if (match == null) {
+      usageException('$package is not a valid package specifier.');
+    }
+
+    final packageName = match.namedGroup('name')!;
+    final descriptor = match.namedGroup('descriptor')?.substring(1);
+
+    /// We want to allow for [constraint] to take on a `null` value here to
+    /// preserve the fact that the user did not specify a constraint.
+    VersionConstraint? constraint;
+
+    /// The package to be added.
+    PackageRef? ref;
+
+    if (descriptor != null) {
+      try {
+        // An unquoted version constraint is not always valid yaml.
+        // But we want to allow it here anyways.
+        constraint = VersionConstraint.parse(descriptor);
+      } on FormatException {
+        final parsedDescriptor = loadYaml(descriptor);
+        // Use the pubspec parsing mechanism for parsing the descriptor.
+        final Pubspec dummyPubspec;
+        try {
+          dummyPubspec = Pubspec.fromMap({
+            'dependencies': {
+              packageName: parsedDescriptor,
+            }
+          }, cache.sources,
+              // Resolve relative paths relative to current, not where the pubspec.yaml is.
+              location: p.toUri(p.join(p.current, 'descriptor')));
```

Comment:
Really? I don' t know the answer, just asking :rofl:

Reply:
I think this is right. We need a `file name`, because the pubspec parsing `location` parameter expects the path to a file name, not the containing directory.